### PR TITLE
feat(api): conf info for upload

### LINKED ISSUE #2504: Conf
Unified Report Settings (GET & SET), SPDX Report Settings (GET & SET), Obligations Settings (GET & SET), Decision Settings (GET & SET)

### PR DESCRIPTION
**Pull request by soham4abc**
## Description

This endpoint provides us full conf information as a response for an Upload. 

## How to test

Send a GET request to `/file/upload/{UploadId}/item/{UploadId}/conf`

## Screenshots
image

Closes #2504

<a href="https://gitpod.io/#https://github.com/fossology/fossology/pull/2505"></a>

### CODE REVIEW

File: src/www/ui/api/Controllers/ConfController.php [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+      if (!$this->dbHelper->doesIdExist("upload", "upload_pk", $uploadPk)) {
+        $returnVal = new Info(404, "Upload does not exist", InfoType::ERROR);
+      } else if (!$this->dbHelper->doesIdExist($this->uploadDao->getUploadtreeTableName($uploadPk), "uploadtree_pk", $uploadTreeId)) {
+        $returnVal = new Info(404, "Item does not exist", InfoType::ERROR);
+      }
+      if ($returnVal !== null) {
+        return $response->withJson($returnVal->getArray(), $returnVal->getCode());
+      }
+      $response_view = $this->uploadDao->getReportInfo($uploadPk);
+      return $response->withJson($response_view, 200);
```
[the remark this answers is not in the document]
Comment (GMishx) [reply src/www/ui/api/Controllers/ConfController.php:83]: Use following JSON format for the same:
```json
{
  "reviewed": "Some Name",
  "footer": "Some Organization",
  "reportRel": "Some release date",
  "community": "Community name",
  "component": "Component name",
  "version": "1.2.3",
  "releaseDate": "12\\/05\\/2018",
  "sw360Link": "NA",
  "componentType": "package-url",
  "componentId": "pkg:npm\\/30-seconds-of-code@1.2.3",
  "generalAssesment": "Some assessment test",
  "gaAdditional": "NA",
  "gaRisk": "No additional information\\/violations found for the package.",
  "gaCheckbox": "unchecked,unchecked,unchecked,unchecked,unchecked,unchecked,unchecked,unchecked,unchecked",
  "spdxSelection": "checked,unchecked",
  "excludedObligations": {
    "Obligation 1": [
      "MIT",
      "GPL-3.0-only"
    ]
  },
  "department": "FOSSology Generation",
  "depNotes": "NA",
  "exportNotes": "NA",
  "copyrightNotes": "NA",
  "unifiedColumns": {"assessment":{"Assessment Summary":"on"},"compliancetasks":{"Required license compliance tasks":"on"},"acknowledgements":{"Acknowledgements":"on"},"exportrestrictions":{"Export Restrictions":"on"},"intellectualProperty":{"Patent Relevant Statements":"on"},"notes":{"Notes":"on"},"scanresults":{"Results of License Scan":"on"},"mainlicenses":{"Main Licenses":"on"},"redlicense":{"Other OSS Licenses (red) - Do not Use Licenses":"on"},"yellowlicense":{"Other OSS Licenses (yellow) - additional obligations to common rules (e.g. copyleft)":"on"},"whitelicense":{"Other OSS Licenses (white) - only common rules":"on"},"overviewwithwithoutobligations":{"Overview of All Licenses with or without Obligations":"on"},"copyrights":{"Copyrights":"on"},"copyrightsuf":{"Copyrights (User Findings)":"on"},"bulkfindings":{"Bulk Findings":"on"},"licensenf":{"Non-Functional Licenses":"on"},"irrelevantfiles":{"Irrelevant Files":"on"},"dnufiles":{"Do not use Files":"on"},"changelog":{"Clearing Protocol Change Log":"on"}},
  "globalDecision": false
}
```

Additional notes:
1. `componentType`: Use map from class \Fossology\Lib\Data\Package\ComponentType
2. `excludedObligations`: Can be an object, `null` or empty array `[]`
3. `unifiedColumns`: It is an object with key: "report conf section id" and value: "{display text, on/off}"
4. `globalDecision`: It is a boolean.

**Review by GMishx — CHANGES_REQUESTED** [review]

**Comment by soham4abc** [comment]
new response format:

image

File: src/www/ui/api/Controllers/ConfController.php [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+  private $uploadDao;
+
+  public function __construct($container)
+  {
+    parent::__construct($container);
+    $this->container = $container;
+    $this->clearingDao = $this->container->get('dao.clearing');
+    $this->licenseDao = $this->container->get('dao.license');
+    $this->uploadDao = $container->get('dao.upload');
+    $this->restHelper = $container->get('helper.restHelper');
```
Comment (GMishx) [inline src/www/ui/api/Controllers/ConfController.php:58]: Container is set in parent class.

```suggestion
    parent::__construct($container);
    $this->clearingDao = $this->container->get('dao.clearing');
    $this->licenseDao = $this->container->get('dao.license');
```

File: src/www/ui/api/Controllers/ConfController.php [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+   */
+  private $uploadDao;
+
+  public function __construct($container)
+  {
+    parent::__construct($container);
+    $this->container = $container;
+    $this->clearingDao = $this->container->get('dao.clearing');
+    $this->licenseDao = $this->container->get('dao.license');
+    $this->uploadDao = $container->get('dao.upload');
```
Comment (GMishx) [inline src/www/ui/api/Controllers/ConfController.php:57]:
```suggestion
    $this->uploadDao = $this->restHelper->getUploadDao();
```

File: src/www/ui/api/documentation/openapi.yaml [hunk 4]
```diff
@@ -4418,6 +4444,97 @@ components:
               enum:
                 - OK
                 - ERROR
+    confResponse:
```
Comment (GMishx) [inline src/www/ui/api/documentation/openapi.yaml:4447]: Please following naming convention
```suggestion
    ConfResponse
```

File: src/www/ui/api/Models/Conf.php [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+      'gaAdditional' => $this->data["ri_ga_additional"],
+      'gaRisk' => $this->data["ri_ga_risk"],
+      'gaCheckbox' => $this->data["ri_ga_checkbox_selection"],
+      'spdxSelection' => $this->data["ri_spdx_selection"],
+      'excludedObligations' => $this->data["ri_excluded_obligations"],
+      'department' => $this->data["ri_department"],
+      'depNotes' => $this->data["ri_depnotes"],
+      'exportNotes' => $this->data["ri_exportnotes"],
+      'copyrightNotes' => $this->data["ri_copyrightnotes"],
+      'unifiedColumns' => $this->data["ri_unifiedcolumns"],
```
Comment (GMishx) [inline src/www/ui/api/Models/Conf.php:77]: Translate this data back as an array. It currently renders as a JSON string.

**Review by GMishx — CHANGES_REQUESTED** [review]

File: src/www/ui/api/Models/Conf.php [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+      'releaseDate' => $this->data["ri_release_date"],
+      'sw360Link' => $this->data["ri_sw360_link"],
+      'componentType' => ComponentType::TYPE_MAP[$this->data["ri_component_type"]],
+      'componentId' => $this->data["ri_component_id"],
+      'generalAssesment' => $this->data["ri_general_assesment"],
+      'gaAdditional' => $this->data["ri_ga_additional"],
+      'gaRisk' => $this->data["ri_ga_risk"],
+      'gaCheckbox' => $this->data["ri_ga_checkbox_selection"],
+      'spdxSelection' => $this->data["ri_spdx_selection"],
+      'excludedObligations' => $this->data["ri_excluded_obligations"],
```
Comment (GMishx) [inline src/www/ui/api/Models/Conf.php:72]: Translate this as well from JSON string to array. See comment https://github.com/fossology/fossology/pull/2505#discussion_r1258024534

**Review by GMishx — CHANGES_REQUESTED** [review]

File: src/www/ui/api/Controllers/ConfController.php [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+
+class ConfController extends RestController
+{
+  /**
+   * @var ContainerInterface $container
+   * Slim container
+   */
+  protected $container;
+
```
Comment (GMishx) [inline src/www/ui/api/Controllers/ConfController.php:33]: No need to re-declare the property again. Please remove.

File: src/www/ui/api/Models/Conf.php (hunk 6 above)
Comment (GMishx) [reply src/www/ui/api/Models/Conf.php:72]: I see no code changes here. Why the conversation was marked as resolved?

File: src/www/ui/api/Models/Conf.php (hunk 5 above)
Comment (GMishx) [reply src/www/ui/api/Models/Conf.php:77]: I see no code changes here. Why the conversation was marked as resolved?

**Review by GMishx — CHANGES_REQUESTED** [review]

**Comment by soham4abc** [comment]
@GMishx I am not able to understand how do I translate the response to an array from JSON string. Can you provide some insight and information regarding this please?

**Comment by GMishx** [comment]
> @GMishx I am not able to understand how do I translate the response to an array from JSON string. Can you provide some insight and information regarding this please?

So please ask question, don't simply mark discussion as resolved.

[json_decode()](https://www.php.net/manual/en/function.json-decode.php) is your friend.

**Comment by soham4abc** [comment]
> 

Sure!!

My bad, I thought having it under array() would resolve the issue but it was a wrong approach.

Thanks for the information!!! 

@GMishx

**Comment by soham4abc** [comment]
> > @GMishx I am not able to understand how do I translate the response to an array from JSON string. Can you provide some insight and information regarding this please?
> 
> So please ask question, don't simply mark discussion as resolved.
> 
> [json_decode()](https://www.php.net/manual/en/function.json-decode.php) is your friend.

I tried to use json_decode() method but it says it's already an array.

image

And the returned result is `NULL`.

**Comment by GMishx** [comment]
> > > @GMishx I am not able to understand how do I translate the response to an array from JSON string. Can you provide some insight and information regarding this please?
> > 
> > 
> > So please ask question, don't simply mark discussion as resolved.
> > [json_decode()](https://www.php.net/manual/en/function.json-decode.php) is your friend.
> 
> I tried to use json_decode() method but it says it's already an array.
> 
> image
> 
> And the returned result is `NULL`.

Ummm, why are you decoding an array?

I expected to decode just the `ri_excluded_obligations` and `ri_unifiedcolumns`.

**Comment by soham4abc** [comment]
> > > > @GMishx I am not able to understand how do I translate the response to an array from JSON string. Can you provide some insight and information regarding this please?
> > > 
> > > 
> > > So please ask question, don't simply mark discussion as resolved.
> > > [json_decode()](https://www.php.net/manual/en/function.json-decode.php) is your friend.
> > 
> > 
> > I tried to use json_decode() method but it says it's already an array.
> > image
> > And the returned result is `NULL`.
> 
> Ummm, why are you decoding an array?
> 
> I expected to decode just the `ri_excluded_obligations` and `ri_unifiedcolumns`.

I'm, Sorry I completely misunderstood the requested change.

Made the changes and pushed as of now

The current response:

image

File: src/www/ui/api/Controllers/ConfController.php (hunk 7 above)
Comment (GMishx) [reply src/www/ui/api/Controllers/ConfController.php:33]: Reminder @soham4abc

File: src/www/ui/api/Controllers/ConfController.php [hunk 8]
```diff
@@ -0,0 +1,56 @@
+<?php
+/*
+ SPDX-FileCopyrightText: © 2018 Siemens AG
```
Comment (GMishx) [inline src/www/ui/api/Controllers/ConfController.php:3]: Please remove invalid copyright text.

File: src/www/ui/api/Controllers/ConfController.php (hunk 7 above)
Comment (GMishx) [inline src/www/ui/api/Controllers/ConfController.php:33]: Please remove as requested.

File: src/www/ui/api/documentation/openapi.yaml [hunk 9]
```diff
@@ -4987,6 +5013,97 @@ components:
...
+        unifiedColumns:
+          type: object
+          properties:
+            report conf section id:
+              type: string
+              enum:
+                - display text
+                - on
+                - off
+              example: display text
```
Comment (GMishx) [inline src/www/ui/api/documentation/openapi.yaml:5095]:
```suggestion
                - on
                - off
              example: on
```

**Review by GMishx — CHANGES_REQUESTED** [review]

**Comment by GMishx** [comment]
Tested, working as expected.

**Comment by soham4abc** [comment]
Made the changes as per request. 
Please do have a look!!

Thank you!

**Review by GMishx — APPROVED** [review]
Changes looks good.

**Pull request merged**